### DISCUSSION TITLE
Xamarin.Forms/MAUI to Avalonia

### DISCUSSION THREAD

**Discussion by marcosgerene** [AvaloniaUI/Avalonia#14842] [Q&A] [answered]
Hi folks, how are you doing?

I have an Xamarin.Forms project in production for some years. It is posted on Android and iOS stores. It is not "public" application, it is part of my ERP for mobile use.

I tried to migrate to MAUI, it was disaster, MAUI is far way to be ready for production. All my base code as models, view models are in dotNet, so stay with dotNet is by far my first choice.

My question is, thinking just about iOS and Android (smartphones), how mature is Avalonia for "real" world. We lost some weeks on MAUI, and we just gave up. Should I try Avalonia as solution to stay with dotNET?

How does it compares to UNO? (https://platform.uno/ios-and-android/)

I have plans to migrate from WPF to Avalonia on Desktop and bigscreens, as I said here: https://github.com/AvaloniaUI/Avalonia/discussions/14188

My question now is I have to get out of Xamarin.Forms ASAP and MAUI was a big disappointment.

Thank you.

**Comment by timunie** [comment]
@marcosgerene we have some Android Apps on the google store for example, so ppl using it in production already. It gets better every day from my point of view. So I'd suggest you do a POC first to figure out all your needs are met. Other than UNO, all controls look the same on all platforms. That's the main difference afaik. 

Stunning sample: https://github.com/AvaloniaUI/Solitaire

**Comment by jonmdev** [comment]
I am looking at the exact same problem. Only I spent months on Maui. I have submitted at least 20-30 Maui bug reports over the past 6 months of which almost all have been ignored, even though they are of major issues like Labels truncating entire letters/words in Android.

The Editor (text field) on iOS is still broken 5-6 months after I reported it as well though they are working on it (should it take 5-6+ months for something so important?)

Microsoft clearly does not care about Maui as they are fixing bugs so slowly it will take 1-2+ years more probably before it is usable.

I find native development is an absolute pain in the ass as well. I am considering Avalonia as well for this reason to stay in .NET. Most of my custom code should be able to move over this way.

**Reply by marcosgerene** [reply]
I had a call with a guy from Uno in Brazil... I will give a look at it also.

MAUI is dead, it is a matter of time.

**Reply by jonmdev** [reply]
Perhaps. They are still working on it. Once it becomes stable it will not need major maintenance. In 1-2 years at this rate it should be pretty stable. The Windows/iOS/Android OS systems themselves don't change over time after that in any significant way.

It is hard to find any good system that does everything well.

Everything seems to have shortcomings. Like with Avalonia I see no good integrated video player that works well in every OS and can allow overlying UI text/graphics easily.

It seems we suffer everywhere we go.

**Reply by marcosgerene** [reply]
@jonmdev 

I hear you, but I don't think I have the same feeling...

>Perhaps. They are still working on it. Once it becomes stable it will not need major maintenance. In 1-2 years at this rate it should be pretty stable. The Windows/iOS/Android OS systems themselves don't change over time after that in any significant way.

Xamarin.Forms EOL is on May, 2024. It is just now! Could a real-world application "wait" two years to be stable?!

I am sorry, but you can see things like these:

https://github.com/dotnet/maui/discussions/19029
https://github.com/dotnet/maui/discussions/10596

MAUI **needs** Microsoft attention and it is not happening at all, it seems to be an abandoned project!

Avalonia/Uno have to improve in some areas maybe, but it is clear the difference when you are asking of commitment.

> It is hard to find any good system that does everything well.
>
>Everything seems to have shortcomings. Like with Avalonia I see no good integrated video player that works well in every OS and can allow overlying UI text/graphics easily.
>
>It seems we suffer everywhere we go.

You can't use a screwdriver as a hammer. In my case Avalonia is my 1st choice for Desktop. 
For Mobile I came here to ask because my gut says UNO is more mature on iOS/Android development.


MAUI is a M$FT project, not a community one. If they don't care, nobody will. It is dead.

PS: Should I start every sentence with "IMHO"? xD


----

A little more:

https://github.com/dotnet/maui/discussions/15203
https://github.com/dotnet/maui/discussions/11568
https://github.com/dotnet/maui/discussions/8652

**Reply by jp2masa** [reply]
> Like with Avalonia I see no good integrated video player that works well in every OS and can allow overlying UI text/graphics easily.

@jonmdev Although some major controls are currently missing (typically provided by third parties), the cool thing about Avalonia is that you can implement almost anything you want without forking, even if you need lower-level APIs. For example, for video players, you can use the GPU interop feature (#10014, not sure if there is any sample besides [samples/GpuInterop](https://github.com/AvaloniaUI/Avalonia/tree/master/samples/GpuInterop)). You should be able to overlay controls this way.

**Comment by orochasamuel** [comment]
Hey @marcosgerene, what's up?

Long lecture but, is pure gold. See below:

https://www.codeproject.com/Articles/5366945/Multiplatform-XAML-Csharp-Miracle-Package-Avalonia
https://github.com/robloo/PublicDocs/blob/master/XAMLFrameworkComparison.md

Live long and prosper 🖖

**Reply by marcosgerene** [reply, marked as answer]
@orochasamuel 
Thanks!

TL;DR;
We decides by Avalonia, but I want to add some context here.


---

This app I mentioned is a small part of my main application, so I could delegate to another dev. So, everything I will say now could be biased in his opinion, even with all my research.


**MAUI** is not ready. Even worse, it is broken.
I talked to a collegue and he defended it in some points, but long story short, he uses SyncFusion. Maui by itself does not work. We had to add a lot of workaround but new versions brought new problems, then we gave up.

**UNO** I had a call with a guy from Brazil who was so helpful. Before that I have not considered UNO in a serious way. Going to tests, our dev reported some little things bothering him during the small experience. Eg he couldn't use dot (.) in the project's name, like _Project.Mobile.Uno_ (to be fair, I did not try myself).
Once we lost some weeks on MAUI we decided to not keep trying with UNO. It is a pretty small project, but time is money.

**Avalonia** seems to be easier. It just works. 3rd part components/themes as Material, Semi, Ursa are just ready to use.
You can find solution to a lot of things as Shell navigation tottaly open source and somehow documented.


For those who are reading this in future trying to get out from Xamarin Forms: Avalonia works.

**Reply by timunie** [reply]
Just to add, in case someone needs priority fixes and or faster support: We do offer payed support for commercial Apps.

**Reply by danipen** [reply]
> 3rd part components/themes as Material, Semi, Ursa are just ready to use.

Interesting to checkout too just in case you don't know them...
https://www.actiprosoftware.com/products/controls/avalonia
